subprojects/enterprise-operations/src/main/java/org/gradle/operations/dependencies/transforms/SnapshotTransformInputsBuildOperationType.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.operations.dependencies.transforms;
+
+import org.gradle.api.internal.tasks.SnapshotTaskInputsBuildOperationType;
+import org.gradle.internal.operations.BuildOperationType;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents the computation of capturing the before execution state and resolving the caching state for transforms.
+ * <p>
+ * Must occur as a child of {@link org.gradle.operations.execution.ExecuteWorkBuildOperationType}.
+ *
+ * @since 8.3
+ */
+public final class SnapshotTransformInputsBuildOperationType implements BuildOperationType<SnapshotTransformInputsBuildOperationType.Details, SnapshotTransformInputsBuildOperationType.Result> {
+    public interface Details {
+    }
+
+    /**
+     * The hashes of the inputs.
+     * <p>
+     * If the inputs were not snapshotted, all fields are null.
+     */
+    public interface Result {
+
+        /**
+         * The overall hash value for the inputs.
+         * <p>
+         * Null if the overall key was not calculated because the inputs were invalid.
+         */
+        @Nullable
+        byte[] getHashBytes();
+
+        /**
+         * The hash of the classloader that loaded the transform action implementation.
+         * <p>
+         * Null if the classloader is not managed by Gradle.
+         */
+        @Nullable
+        byte[] getClassLoaderHashBytes();
+
+        /**
+         * The class name of the transform's action.
+         */
+        @Nullable
+        String getImplementationClassName();
+
+        /**
+         * The input value property hashes.
+         * <p>
+         * key = property name
+         * <p>
+         * Ordered by key, lexicographically.
+         * No null keys or values.
+         * Never empty.
+         */
+        @Nullable
+        Map<String, byte[]> getInputValueHashesBytes();
```

Comment:
🤔 Hash codes are used all over the place. This seems like an opportunity to introduce a type for them. One reason to do so would be that with the new cache protocol we might want to switch to "digests", i.e. a record that keeps track of the hash _and_ the length of the data. It would make it easier to expose that data if we had a richer type instead of `byte[]`.

Reply:
Capturing the full digest would allow input comparisons to show differences in size when there's a mismatch. Another field we might end up adding to digests is the hash algorithm used.

Reply:
I don't see how the size helps in the comparison: if the size changed, the contents changed as well. I don't see how it is helpful to show that the size stayed the same and only the contents changed.

Reply:
We discussed this and another reason we might want to do this is to avoid creating garbage by creating these byte[]. Gradle's `HashCode` now uses two `long`s to store the hash code, and we could do something similar in a type for build operations. Though we'd do this change for all operations not only for transforms. Let's keep this a `byte[]` for now.